### ISSUE TITLE
UseCheckOrError/UseRequire false positives

### ISSUE THREAD

**Issue by technoir42** [arturbosch/detekt#1683]
detekt 1.0.0-RC15

These false positives can be reported by either UseCheckOrError or UseRequire depending on the type of exception (IllegalStateException/IllegalArgumentException).

**Case 1: when**
```kotlin
override fun getColumnCount(item: IndexedItem): Int = when (item) {
    is FolderItem -> folderParams.columnCount
    is ImageItem -> imageParams.columnCount
    else -> throw IllegalArgumentException("Unexpected item ${item.javaClass}")
}
```

**Case 2: no condition around throw**
```kotlin
fun getTagIndex(tag: Tag, tags: List<Tag>): Int {
    var index = 0
    for (t in tags) {
        if (t === tag) {
            return index
        } else if (t.type == tag.type) {
            index++
        }
    }
    throw IllegalArgumentException("Tag $tag was not found in the list") // UseRequire
}
```

**Case 3: elvis operator**
```kotlin
private fun getSuperclassTypeParameter(subclass: Class<*>): Type {
    val superclass = subclass.genericSuperclass as? ParameterizedType
        ?: throw IllegalArgumentException("Super class of $subclass must be parametrized") // UseRequire
    return superclass.actualTypeArguments[0]
}
```

**Comment by cortinico** [comment]
Thanks for the report @technoir42 
Here a potential fix to the issue you reported #1689

**Comment by arturbosch** [comment]
I see, so basically the problem is that `require` and `check`do not return `Nothing` and can't be used as expressions. However `error` can be used as an expression.
I agree with your changes @cortinico for `UseRequire` and `UseCheck`. Maybe we should introduce a new rule `UseError` with the old behavior?
